### PR TITLE
Fixes for C++17, undefined behavior, and compiler warnings

### PR DESCRIPTION
**Pull request by davidrohr**
Larg commit with fixes for C++17 with strict compilers (getting rid of std::auto_ptr in particular), in parallel, I this cleans up a large fraction of the compiler warnings because I was sick of it.

We should definitely run this through release validation first (@dberzano).

- Remove register keyword
- Use AUTO_PTR macro to switch between auto_ptr and unique_ptr for C++98/C++17
- Change all ROOT_INCLUDE_PATHS to system includes to suppress warnings from ROOT headers.
- Fix plenty of compiler warnings, add missing ROOT include PATHS as SYSTEM includes, which worked before for some weird reason.
- Fixed some array out of bounds access
- Changed some return type functions that were returning references to local variables, and where this was not possible, at least mode them static.

@aphecetche : Could you check whether this fixes you compilation issues on Mac with C++17, and if not let me know which errors remain? (I do not have a mac at hand, and the CI will only test on Mac wo C++17)

### CODE REVIEW

**Comment by davidrohr** [comment]
+test

**Comment by aphecetche** [comment]
Still has the alisync errors : 

[log2.txt](https://github.com/alisw/AliRoot/files/2493600/log2.txt)

**Comment by davidrohr** [comment]
@aphecetche : I have updated the PR with more fixes to get rid of the std::ptr_fun, that broke alisync for you. Could you check again if there is more?

**Comment by aphecetche** [comment]
@davidrohr compilation is now going through, yes. 

Note that launching AliRoot I get a strange warning about AliRun

AliRoot/latest-PR804-2-o2] ~/alice/o2-dev %> aliroot -b
input_line_14:2:6: error: unknown type name 'AliRun'
 new AliRun("gAlice", "The ALICE Off-line Simulation Framework");
     ^
   ------------------------------------------------------------
  | Welcome to ROOT 6.14/04                http://root.cern.ch |
  |                               (c) 1995-2018, The ROOT Team |
  | Built for macosx64                                         |
  | From tags/v6-14-04@v6-14-04, Oct 18 2018, 12:35:00         |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q' |
   ------------------------------------------------------------

root [0]

**Comment by davidrohr** [comment]
Weird, I do not see this error message.
This PR does not touch AliRun.h, so perhaps it is unrelated.

@pzhristov @dberzano : Do you have any idea what it could mean.

**Comment by davidrohr** [comment]
+test

**Comment by davidrohr** [comment]
+test

**Comment by davidrohr** [comment]
Hi @dberzano @pzhristov : 

did you have a chance to look at this? What is your general feeling? Would it make sense to go on fixing the warnings that show up very frequently, or do you think we do not want to merge such a big change?

Cheers
David

File: T0/T0calib/AliT0TimeAmplCorr.cxx [hunk 1]
```diff
@@ -400,7 +400,7 @@ Int_t AliT0TimeAmplCorr::SetNewQTCcoefficients(TString filename)
             if((runMax - runMin) < minRUNdiff) enteryChoosen = enteryForMeanValues;
             tree_coef->GetEntry(enteryChoosen);
 
-            for(Int_t pmt = 0; pmt <= NPMT0; pmt++)
```
Comment (dberzano) [inline T0/T0calib/AliT0TimeAmplCorr.cxx]: Are you sure? I mean: it makes sense to me, and I believe you addressed a specific compiler warning here, but we must be sure this really does what we want.

File: ACORDE/ACORDEbase/AliACORDEPreprocessor.cxx [hunk 2]
```diff
@@ -150,9 +150,10 @@ UInt_t AliACORDEPreprocessor::Process(TMap* dcsAliasMap)
         TString fileName = GetFile(kDAQ,SourcesId.Data(), source->GetName());
 
   	if (fileName.Length() > 0)
+  	{
```
Comment (dberzano) [inline ACORDE/ACORDEbase/AliACORDEPreprocessor.cxx]: Here you have tabs instead of spaces I believe (it's misaligned with the closing one).

**Comment by dberzano** [comment]
Hi @davidrohr, as per policy I never take a look at PRs marked as `WIP` 😄 Now I did, see my review.

File: ACORDE/ACORDEbase/AliACORDEPreprocessor.cxx (hunk 2 above)
Comment (davidrohr) [reply ACORDE/ACORDEbase/AliACORDEPreprocessor.cxx]: The indentation in that file is totally messed up, But yes, I will fix the indentation in the part I changed.

File: T0/T0calib/AliT0TimeAmplCorr.cxx (hunk 1 above)
Comment (davidrohr) [reply T0/T0calib/AliT0TimeAmplCorr.cxx]: Yes, I checked that, and actually it caused a buffer overflow before.

**Comment by davidrohr** [comment]
Hi @dberzano : OK, so if you have more comments let me know, and I will address them.
If you like, I will have one more go, and another time for frequent warnings from headers included many times. Fixing them will reduce the number of total errors significantly.

I actually tried to just put -Werror -Wall -Wextra and see how far I get, but after 2 hours I only got as far as 5% of the CMake compilation so I gave up.

I'll update the PR in the next days. To be absolutely sure, I'd suggest to run it through the release validation before we merge it.

Cheers
David

**Comment by dberzano** [comment]
@davidrohr I discourage you from adding `-Werror`. I liked it in the past, but it seems that every new compiler changes what is considered an error, and there are even differences between Clang and GCC requiring to add "no error" exceptions here and there. It's better to have a close look at the warnings instead. O2 is a different story, there it makes sense.

As for validating it, we will merge and release-validate it. If it does not work, we will revert it (the procedure is a bit easier like this)!

Thanks for this PR. If I was there I'd buy you an orange-flavored chocolate icecream at [Mara dei Boschi](http://www.maradeiboschi.com/) 🍦 😄

**Comment by pzhristov** [comment]
Hi @davidrohr ,
I went through the changes, they are OK for me. Peter

**Comment by davidrohr** [comment]
+test

**Comment by davidrohr** [comment]
@dberzano : I fixed the indentation stuff that you commented on (although there is plenty of more broken indentation...), and did one more run to fix most warnings in headers.
So I will not go on here and from my side the PR can be merged, although I would recommend to do a release validation soon after.

**Comment by davidrohr** [comment]
+test (once again after minor fix)

**Comment by dberzano** [comment]
+1 merging this, as we will start a RelVal soon. Will revert in case of troubles.

Just a clarification: even though C++17 would be desirable in some cases, there is no plan to move current Run 2 software further than C++11 as our efforts are concentrated on Run 3 now. Also, the tests we are performing here do not use `-std=c++17`.

**Pull request merged**